blocks/api/post.pegjs

```diff
@@ -1,33 +1,179 @@
+ *                                              __/ |
+ *                  GRAMMAR                    |___/
+ *
+ *
+ * Welcome to the grammar file for Gutenberg posts!
+ *
+ * Please don't be distracted by the functions at the top
+ * here - they're just helpers for the grammar below. We
+ * try to keep them as minimal and simple as possible,
+ * but the parser generator forces us to declare them at
+ * the beginning of the file.
+ *
+ * What follows is the official specification grammar for
+ * documents created or edited in Gutenberg. It starts at
+ * the top-level rule `Block_List`
+ *
+ * The grammar is defined by a series of _rules_ and ways
+ * to return matches on those rules. It's a _PEG_, a
+ * parsing expression grammar, which simply means that for
+ * each of our rules we have a set of sub-rules to match
+ * on and the generated parser will try them in order
+ * until it finds the first match.
+ *
+ * This grammar is a _specification_ (with as little actual
+ * code as we can get away with) which is used by the
+ * parser generator to generate the actual _parser_ which
+ * is used by Gutenberg. We generate two parsers: one in
+ * JavaScript for use the browser and one in PHP for
+ * WordPress itself. PEG parser generators are available
+ * in many languages, though different libraries may require
+ * some translation of this grammar into their syntax.
+ *
+ * For more information:
+ * @see https://pegjs.org
+ * @see https://en.wikipedia.org/wiki/Parsing_expression_grammar
+ *
+ */
+
 /** <?php
 // The `maybeJSON` function is not needed in PHP because its return semantics
 // are the same as `json_decode`
+
+// array arguments are backwards because of PHP
+if ( ! function_exists( 'peg_array_partition' ) ) {
+    function peg_array_partition( $array, callable $predicate ) {
+        $truthy = array();
+        $falsey = array();
+
+        foreach ( $array as $item ) {
+            call_user_func( $predicate, $item )
+                ? $truthy[] = $item
+                : $falsey[] = $item;
+        }
+
+        return array( $truthy, $falsey );
+    }
+}
+
+if ( ! function_exists( 'peg_join_blocks' ) ) {
+    function peg_join_blocks( $pre, $tokens, $post ) {
+        $blocks = array();
+        
+        if ( ! empty( $pre ) ) {
+            $blocks[] = array( 'blockName' => 'core/freeform', 'innerHTML' => $pre );
+        }
+        
+        foreach ( $tokens as $token ) {
+            list( $token, $html ) = $token;
+            
+            $blocks[] = $token;
+            
+            if ( ! empty( $html ) ) {
+                $blocks[] = array( 'blockName' => 'core/freeform', 'innerHTML' => $html );
+            }
+        }
+        
+        if ( ! empty( $post ) ) {
+            $blocks[] = array( 'blockName' => 'core/freeform', 'innerHTML' => $post );
+        }
+        
+        return $blocks;
+    }
+}
+
 ?> **/
 
+function freeform( s ) {
+  return s.length && {
+    blockName: 'core/freeform',
```

Comment:
Elsewhere we have made assumptions that freeform is the absence of a block name. Specifically, [this test case](https://github.com/WordPress/gutenberg/blob/6e984b0a13392bd85159eccc989bd56d4cdd0779/blocks/api/test/parser.js#L381-L416) now fails for a few reasons:

- In block registration we allow unknown type handler to be assigned to any block name, not assuming `core/freeform`
- The content of the freeform block now includes the block delimiters `<!-- wp:freeform -->` which was not the case previously and arguably not desirable (see also #2513)